### ISSUE TITLE
Solaris FQDN/UQDN and documentation/consistancy

### ISSUE THREAD

**Issue by sjorge** [saltstack/salt#37027]
### Description of Issue/Question

Follow up to #36869 

I did some more digging into this, generally speaking...

On linux nodename is set to the UQDN, hostname returns the UQDN unless -f is provided.
On solaris nodename is set to FQDN, hostname returns FQDN unless -s is provided then UQDN.

However historically a lot of /etc/nodename seem to have been set to UQDN (google digs up a lot of these type of results)

This leaves us with a bit of a pickle that probably is best resolved in the documentation.

With /etc/nodename set to a UQDN, socket.getfqdn() just returns the UQDN.
When it is set to FQDN, the proper FQDN is returned.

I quickly grepped the docs and found this on the salt.7 manpage

``` man
.SS \fBappend_domain\fP
.sp
Default: \fBNone\fP
.sp
Append a domain to a hostname in the event that it does not exist. This is
useful for systems where \fBsocket.getfqdn()\fP does not actually result in a
FQDN (for instance, Solaris).
```

Which seems to indeed assume nodename is set to UQDN.

pinging @isbm, does get_fqhostname honer the 'append_domain' parameter?

We should be consistent in how we handle this, we either assume nodename is FQDN or UQDN. Not UQDN in one case and FQDN in another case.

But from my understanding now is that FQDN is the correct answer which is different from linux and friends. The behavior of socket.getfqdn() and the hostname command seems to support this.

I did find check-hostname command that seems to provide the correct answer for all my minions

```
CHECK-HOSTNAME(1M)           Maintenance Commands           CHECK-HOSTNAME(1M)



NAME
       check-hostname - check if sendmail can determine the system's fully-
       qualified host name

SYNOPSIS
        /usr/sbin/check-hostname


DESCRIPTION
       The check-hostname script is a migration aid for sendmail(1M). This
       script tries to determine the local host's fully-qualified host name
       (FQHN) in a manner similar to sendmail(1M). If check-hostname is able
       to determine the FQHN of the local host, it reports success. Otherwise,
       check-hostname reports how to reconfigure the system so that the FQHN
       can be properly determined.

FILES
       /etc/hosts
                             Host name database


       /etc/nsswitch.conf
                             Name service switch configuration file


       /etc/resolv.conf
                             Configuration file for name server routines


ATTRIBUTES
       See attributes(5) for descriptions of the following attributes:




       +--------------------+-----------------+
       |  ATTRIBUTE TYPE    | ATTRIBUTE VALUE |
       +--------------------+-----------------+
       |Interface Stability | Evolving        |
       +--------------------+-----------------+

SEE ALSO
       domainname(1M), sendmail(1M), hosts(4), attributes(5)



                                 March 8, 2006              CHECK-HOSTNAME(1M)
```

``` bash
#!/bin/sh --
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

# Check hostname configuration as per the sendmail code.
#
# See http://www.sendmail.org/sun-specific/migration.html#FQHN for details.
#
# Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
# Use is subject to license terms.
#

PATH=/bin:/usr/sbin

# If $1 has a ".", accept it and exit.

accept_if_fully_qualified() {
        case $1 in
        *.*)
                echo "Hostname $myhostname OK: fully qualified as $1"
                exit 0
                ;;
        esac
}

# Check the `getent hosts $1` output, skipping the 1st entry (IP address).

check_gethostbyname() {
        for host in `getent hosts $1 | awk '{for (f=2; f <= NF; f++) print $f}'`
        do
                accept_if_fully_qualified $host
        done
}

# Parse /etc/hosts, looking for $1 as an entry by itself, and try to find
# a long name on the same line.  First kill all comments, then check for
# $1 as a word by itself, then take just the first such line, then skip
# its first entry (IP address).

check_hosts_file() {
        for entry in `sed -e 's/#.*$//' /etc/hosts | \
                awk '/[         ]'$1'([         ]|$)/ \
                        {for (f=2; f <= NF; f++) print $f; exit}'`
        do
                accept_if_fully_qualified $entry
        done
}

# Parse the output of `nslookup $1`, checking the Name and Aliases.

check_dns() {
        for host in `nslookup $1 2>/dev/null | \
                awk '$1 == "Name:" || $1 == "Aliases:"{print $2}'`
        do
                accept_if_fully_qualified $host
        done
}

# Check the `ypmatch $1 hosts` output, skipping the 1st entry (IP address).

check_nis() {
        for hst in `ypmatch $1 hosts | awk '{for (f=2; f <= NF; f++) print $f}'`
        do
                accept_if_fully_qualified $hst
        done
}

# Recommend how to reconfigure to get $1.$2 as the FQHN.
# $3 is the first entry for hosts in /etc/nsswitch.conf .

suggest_fix_and_exit() {
        myhost=$1
        suggested_domain=$2
        fhe=$3
        myipaddr=`getent hosts $myhost | head -1 | awk '{print $1}'`

        # aliases: skip the 1st & 2nd entries: IP address & canonical name

        set -- '' '' '[ aliases ... ]'
        set -- `grep "^$myipaddr[        ]" /etc/hosts 2>/dev/null`
        result=$?
        shift 2
        echo "We recommend \c"
        if [ "x$fhe" != "xfiles" ] ; then
                echo "listing files first for hosts in /etc/nsswitch.conf"
                echo "and then \c"
        fi
        if [ $result = 0 ] ; then
                echo "changing the /etc/hosts entry:\n"
                echo "$myipaddr $myhost $*\n"
                echo "to:\n"
        else
                echo "adding the /etc/hosts entry:\n"
        fi
        echo "$myipaddr $myhost $myhost.$suggested_domain $*"
        exit 0
}

# Fall back to the NIS domain, minus the first label.  If it is non-null,
# use it but recommend against it.  $2 is just informative, indicating whether
# we're checking the NIS domain.  $3 is to pass on.

check_nis_domain() {
        nisdomain=`domainname`
        realdomain=`echo $nisdomain | sed 's/[^.]*\.//'`
        if [ "x$realdomain" != "x" ] ; then
                echo "Hostname $1 can be fully qualified using NIS$2 domain"
                echo "  $nisdomain"
                echo "resulting in the name"
                echo "  $1.$realdomain"
                echo "but this is bad practice.\n"
                suggest_fix_and_exit $1 $realdomain $3
        fi
}

# Goal: try to fully qualify `hostname` as sendmail would.
# Algorithm (stop as soon as a name with a dot is found):
#    1. gethostbyname (simulate with getent hosts)
#    2. fall back to individual hosts: methods in nsswitch.conf, using
#       only those that are configured, in their configured order
#       * files (parse /etc/hosts directly)
#       * dns (parse nslookup output)
#       * nis (parse ypmatch output)
#    3. fall back to the NIS domain name.
# If none of the above succeed, give up.  Recommend:
#    a. the domain entry in /etc/resolv.conf, if one exists
#    b. "pick.some.domain"

myhostname=`hostname`

check_gethostbyname $myhostname

hosts_line=`sed -n -e 's/^hosts:\([^#]*\).*/\1/p' /etc/nsswitch.conf`
first_hosts_entry=`echo $hosts_line | awk '{print $1}'`
nis_domains=""

for entry in $hosts_line
do
        case $entry in
        files)
                check_hosts_file $myhostname
                ;;
        dns)
                check_dns $myhostname
                ;;
        nis)
                check_nis $myhostname
                nis_domains="$nis_domains nis"
                ;;
        esac
done

for entry in $nis_domains
do
        case $entry in
        nis)
                check_nis_domain $myhostname "" $first_hosts_entry
                ;;
        esac
done

realdomain=`awk '$1 ~ /^domain/ {print $2}' 2>/dev/null < /etc/resolv.conf`
case $realdomain in
*.*)
        # OK
        ;;
*)
        realdomain="pick.some.domain"
        ;;
esac

echo "Hostname $myhostname could not be fully qualified."
suggest_fix_and_exit $myhostname $realdomain $first_hosts_entry
```

**Comment by sjorge** [comment]
I switched all my minions to have FQDN in /etc/nodename and the grains are now correct, also on 2016.3.x. FWIW

**Comment by isbm** [comment]
Problem is that Linux doesn't have something like nodename concept in Solaris, if I understand correctly. While `uname -n` is a nodename, and essentially when you call `hostname`, it actually calls `uname -n`. In Solaris a nodename is not really a network name. It just happens that it usually is. But you might change it to whatever you like, like "Umbrella Corp" if your machine has multiple DNS names.

**Comment by damon-atkins** [comment]
https://docs.oracle.com/cd/E18752_01/html/816-4554/ipconfig-12.html

Solaris 10
I would look in /etc/nodename  and if the hostname contain dot, assume it is FQDN and stop
If /etc/nodename  does not contain any dots Then look in /etc/defaultdomain 

Solaris 11
svcprop -p config/nodename svc:/system/identity:node and if the hostname contain dot, assume it is FQDN and stop
Otherwise run domainname and then try svcprop -p config/domain svc:/network/dns/client

**If the nodename is not found in DNS or host file. Solaris boots very slowly (have not check this for a long time)**

I would have thought 
uname -n  and if no dots in the name, then search for the domain.

Last time this came up in general discussion I suggest that the reverse lookup the IP address used to connect to the master.

**Comment by sjorge** [comment]
Sounds reasonable, if that happened my setup would not have broken on 2016.11.0rc1.
As I had UQDN in /etc/nodename and DN in /etc/default/domain

**Comment by Ch3LL** [comment]
I would like to get some others input into this issue to ensure we go forward with a consistent fix. 

ping @cachedout or @rallytime do either of you have any opinions on how to approach this issue?

**Comment by sjorge** [comment]
fwiw, I reverted my switching to having FQDN in /etc/nodename.
It caused some minor issues with other software.

For now I stopped using the FQDN grain.

**Comment by sjorge** [comment]
Bump, I am working around it at the moment by calculating the fqdn grains in a pillar first. (host + '.' + dns:domain) because as mentioned above putting FQDN did cause some issues with other software like postfix going for host.domain.domain instead.

**Comment by damon-atkins** [comment]
Re: Postfix 
I would have thought postfix would not add the domain unless configured too.

I suggest you remove /etc/defaultdomain if you set the hostname to be FQDN
/etc/defaultdomain is meant to be for NIS

Just guessing
It sounds like configs have been setup on the assumption that hostname is not FQDN and hence configs add the domain.

**Comment by sjorge** [comment]
It seems like a bit of a cluster fuck due to legacy reasons. Some take nodename + defaultdomain, some take whatever is in /etc/hosts,...

I already added 'myhostname = correct.fqdn' to the main.cf for postfix, to avoid stuff in the future though.

Let me tag some of the illumos people and see if any of them respond:
@rmustacc, @danmcd can any of you clear this up for illumos/Solaris 10 at least.

Basically, what goes in /etc/nodename, /etc/defaultdomain and /etc/hosts to properly setup a FQDN?

For example of FQDN = mynode.example.org and e1000g0 has ip 192.168.1.10.

```
/etc/nodename:
mynode

/etc/defaultdomain:
example.org

/etc/hosts: 
::1             localhost
127.0.0.1       localhost localhost.local loghost
192.168.1.10 mynode.example.org mynode
```

?

**Comment by isbm** [comment]
@sjorge Why they say `/etc/defaultdomain` is for NIS? It is (not true)[https://docs.oracle.com/cd/E19455-01/806-0916/ipconfig-16/index.html].

**Comment by sjorge** [comment]
@isbm I never said ```/etc/defaultdomain``` was for NIS

But reread the docs and manages doe seem to indicate in my above example I have the correct content in both files.

The facts that the docs uses "ully qualified domain name of the administrative domain" does not really help for clarity but the referenced figure 6-1 is helpful.

**Comment by damon-atkins** [comment]
From the man page for defaultdomain which doco's NIS service generates the /etc/defaultdomain file for backwards compatibility application design to work with older version of Solaris.
```
NAME
     defaultdomain - specify host's domain name

SYNOPSIS
     svc:/network/nis/domain

     /etc/defaultdomain

DESCRIPTION
     The  service  svc:/network/nis/domain  determines  a  host's  domain name for direct
     use by the NIS name service.  For the  purposes of backwards compatibility, the service
     also  generates the defaultdomain file.
```

See also https://docs.oracle.com/cd/E19455-01/806-1387/abtrbl-28007/index.html which talks about fixing NIS.

```
$ svcprop  svc:/network/dns/client/:properties/config/domain
$ svcprop   svc:/network/nis/domain/:properties/config/domainname

```

**Comment by isbm** [comment]
(Default domain file)[https://docs.oracle.com/cd/E19455-01/806-0916/ipconfig-16/index.html] contains an FQDN and is because of (`/etc/hostname.interface`)[https://docs.oracle.com/cd/E19455-01/806-0916/ipconfig-13/index.html]. Same to (IPv6)[https://docs.oracle.com/cd/E19455-01/806-0916/ipref-10/index.html]. AFAIK, (`/etc/nodename`)[https://docs.oracle.com/cd/E19455-01/806-0916/ipconfig-15a/index.html] is not really a FQDN, but just a _machine_ name, not even necessarily resolvable.

Seems pretty clear to me.

**Comment by sjorge** [comment]
Alright lets look at [Figure 6-1](https://docs.oracle.com/cd/E19455-01/806-0916/6ja853993/index.html#ipconfig-fig-10) and create all these files as per the docs for timbuktu and for sahara so we covered a multi homed and singled homed machine.

From how I understand they should look like the following. @isbm, @damon-atkins  do guys come to the same conclusion?

## sahara
/etc/nodename
```
sahara
```

/etc/defaultdomain
```
deserts.worldwide.com
```

/etc/hostname.le0
```
sahara
```

/etc/hosts
```
127.0.0.1    localhost localhost.local loghost
::1          localhost localhost.local loghost
192.9.200.50 sahara.deserts.worldwide.com sahara
```

## timbuktu
/etc/nodename
```
timbuktu
```

/etc/defaultdomain
```
deserts.worldwide.com
```

/etc/hostname.le0
```
timbuktu
```

/etc/hostname.le1
```
timbuktu-201
```

/etc/hosts
```
127.0.0.1    localhost localhost.local loghost
::1          localhost localhost.local loghost
192.9.200.70 timbuktu.deserts.worldwide.com timbuktu
192.9.201.10 timbuktu-201.deserts.worldwide.com timbuktu-201
```

**Comment by damon-atkins** [comment]
Hi @isbm and @sjorge The doco about /etc/defaultdomain is related to NIS.  Most of Sun doco was written when they expected everyone to use NIS.   NIS would not work if /etc/nodename or hostname was FQDN

Solaris 10 and older versions
`/etc/hostname.<interface>` contains options for ifconfig  for a interface. It can contain a hostname or IP address
e.g. From http://docs.oracle.com/cd/E19253-01/816-4554/deploynetmult-56/index.html
```
192.168.85.19 netmask + broadcast + group testgroup1 up \
	addif 192.168.85.21 deprecated -failover netmask + broadcast + up
```
Also available  `/etc/dhcp.<interface>` to trigger dhcp on an interface. If I remember correctly then `/etc/hostname.<interface>` does not or should not exist.

These files do not exist in Solaris 11.

For Solaris 10 the installer does
```
/etc/hosts
IP <same as in /etc/nodename>
IP <FQDN> <ALIAS>
```

For Solaris 11 the installer does, it also has a complete different way to set IP address.
```
/etc/hosts
::1  <svc node name> localhost
IP <FQDN> <ALIAS>
```
In general some programs also look etc /etc/resolv.conf and search for the domain key word to work out the domain.  Hopefully only when the hostname/uname -n is not FQDN.

@sjorge unfortunately can not spend anymore time on this. Try out my suggest and see how you go.

Your friend is `truss -fea <cmd string>`.  He/She will always tell you the truth.
Try truss with `postfix check`   `postconf -d`  and  `postconf -n`    and thing else which has no domain or double domain.

**Comment by sjorge** [comment]
Yeah dhcp.iface en addrconf.iface both also work on Solaris 10 at least.

I doubt we can get this fixed for 2016.11 which basically means that release will break stuff on Solaris. I guess that is unavoidable, but this should be sorted and properly fixed for Nitrogen at the very least. I'm OK in doing the lift work for this, but we need to figure out what is the proper way to get this data that works on both Solaris 10 and friends (illumos/omnios/smartos) and Solaris 11.

I guess re-implementing  fqhostname for SunOS to do what ```check-hostname``` would be a option tool.

Since I am already on 2016.11 I moved all my config away from using the host/fqdn/dns related grains (which is a PITA). Having the stuff configure as #issuecomment-255584442 makes nearly all software do the correct thing, aside from Salt 2016.11, 2016.3 worked fine like that also.

**Comment by damon-atkins** [comment]
Try truss, as every program does its own thing, esp. ones which run on multiple OS.  It's the only way to know for sure, and work out a final solution.

**Comment by sjorge** [comment]
It's been traced to get_fqhostname in salt.utils.network returning different output compared to < 2016.11.

More specifically ``socket.getfqdn()`` and ``socket.getaddrinfo()`` seems to not properly resolve to the fqdn, the only way to force this is to set /etc/nodename to FQDN.

```
 socket.getfqdn([name])

    Return a fully qualified domain name for name. If name is omitted or empty, it is interpreted as the local host. To find the fully qualified name, the hostname returned by gethostbyaddr() is checked, followed by aliases for the host, if available. The first name which includes a period is selected. In case no fully qualified domain name is available, the hostname as returned by gethostname() is returned.
```
```
>>> socket.getfqdn()
'isotope'
>>> socket.gethostbyname(socket.gethostname())
'172.16.10.135'
>>> socket.gethostbyaddr('172.16.10.135')
('isotope.vlan10.acheron.be', ['isotope.acheron.be', 'isotope'], ['172.16.10.135'])
```

After digging into the python source
```
# name is None so gethostname
>>> socket.gethostname()
'isotope'
>>> socket.gethostbyaddr('isotope')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.herror: [Errno 0] host not found
# we end up in the except block, so just name is returned
```

Since both gethostname and gethostbyaddr come form the cbindings, python does not seem to be in the wrong here. Neither does the config (as I understand it) based on the docs.

Previous version worked around this by doing some extra filtering. Since the filtering is kind of horrible, having a Solaris specific implementation of get_fqhostname would solve it, the question is what the proper implementation is.

Or well replacing ``socket.getfqdn()`` by ``socket.gethostbyaddr(socket.gethostbyname(socket.gethostname()))[0]`` seems to give the same result as getfqdn() on Solaris, Linux and OpenBSD (only stuff I tested it on)

**Comment by sjorge** [comment]
Lets summarize that: 172.16.10.135
* Solaris docs are a bit confusing at times
* My config seems to be fine for Solaris 10 and friends at least
* python's socket.getfqdn() returns the wrong thing
* traced to socket.gethostbyaddr(xxx) not returning results on Solaris 10-like systems
  * socket.gethostbyaddr() seems to want a IPv4 or IPv6 address, getfqdn is feeding it a host

**Comment by damon-atkins** [comment]
with nodename/hostname/uname -n  of the system FQDN.
Python  socket.getfqdn() returns FQDN.

Solaris 11 Calls
sysinfo(SI_HOSTNAME, buffer, 1023) and libc:gethostname()

Read the last sentence of below from Solars 11 man page
```
     SI_HOSTNAME

         Copy into the array pointed to  by  buf  a  string  that names  the present host
        machine. This is the string that would be returned by uname() in the
        nodenamefield.  This hostname  or  nodename  is often the name the machine is
        known by locally. The  hostname  is  the  name  of  this machine  as  a  node in
        some network. Different networks might have different names for the node, but
       presenting the  nodename  to  the  appropriate network directory or name-to-address
       mapping service should produce  a  transport  end  point  address.  The name might
       not be fully qualified. Internet host names can be up to 256 bytes in length (plus the
       terminating null).
```
Note:
- "transport  end  point  address" e.g. IP Address
- "The name might not be fully qualified."   So one could read this as expect to be FQDN, but may not be.

Your fix looks up the host file or DNS. And as your host file has the short name as an alias it works. Or resolve.conf has a domain or search.    If you remove resolv.conf and remove the host file line ?

**Comment by sjorge** [comment]
With nothing in /etc/hosts or /etc/resolv.conf

```
>>> socket.gethostbyaddr(socket.gethostbyname(socket.gethostname()))[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.gaierror: [Errno 8] node name or service name not known
```

That being said, I get a similar result on a debian box
```
import Python 2.7.3 (default, Mar 18 2014, 05:13:23)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> socket.gethostbyaddr(socket.gethostbyname(socket.gethostname()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.gaierror: [Errno -3] Temporary failure in name resolution
```

Without /etc/hosts and/or resolve.conf the fqdn grains is not a FQDN on linux either, it is also just the nodename in that case! So the change would not break anything there.

I did dig into the socket library from python, before proposing the above fix,
socket.getfqdn() calls socket.gethostname(), then uses socket.gethostbyaddr() to translate the result into a FQDN.

socket.gethostname() returns just the nodename on my test debian server too (did not set a doamin), but socket.gethostbyaddr() is happy to take the nodename and resolve it into a FQDN, while this on Solaris fails as socket.gethostbyaddr() only wants an IPv4 or IPv6 address.

Without /etc/hosts and/or resolve.conf  socket.gethostbyaddr() will fail on linux, openbsd, and solaris for me. Which seems to suppor that my fix above is indeed correct.

**Comment by sjorge** [comment]
Bump, I have been running with the fix from #issuecomment-255592306 for a few days now and all looks fine with this on both linux and solaris-like minions.

**Comment by thatch45** [comment]
@isbm any issues with the current implementation?

**Comment by damon-atkins** [comment]
This looks like a python dependent version/platform port issue.   I suspect at some stage someone has added code to python so gethostbyaddr() excepts not just an addr but also wrongly accepts a hostname.

I think we should assume gethostbyaddr() does not support a hostname for any platform or any version of python, and follow what is expect by the underlining C code.  Which would make the code more generic. It might be more code, but it will work everywhere, so that their is no need for special cases.

From Fedora
```
       The  gethostbyaddr()  function  returns a structure of type hostent for the given host
       address addr of length len and address type type.  Valid address types are AF_INET
       and AF_INET6.  The host address argument is a pointer to a struct of a type
       depending on the address type, for  example  a  struct in_addr * (probably obtained
       via a call to inet_addr(3)) for address type AF_INET.
```


```
-- Test 1 --
host=google-public-dns-a.google.com
-- Test 2 --
Official name is: www.google.com
IP address: 150.101.161.174
All addresses: 150.101.161.174 150.101.161.180 150.101.161.181 150.101.161.187 150.101.161.146 150.101.161.152 150.101.161.153 150.101.161.159 150.101.161.160 150.101.161.166 150.101.161.167 150.101.161.173
-- Test 3 --
Host name: google-public-dns-a.google.com
[root@master tmp]# cat p.c
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>

int main()
{
    struct sockaddr_in sa;    /* input */
    socklen_t len;         /* input */
    char hbuf[NI_MAXHOST];

    int i;
    struct hostent *he;
    struct in_addr **addr_list;
    struct in_addr addr;

    memset(&sa, 0, sizeof(struct sockaddr_in));

    printf("-- Test 1 --\n");
    /* For IPv4*/
    sa.sin_family = AF_INET;
    sa.sin_addr.s_addr = inet_addr("8.8.8.8");
    len = sizeof(struct sockaddr_in);

    if (getnameinfo((struct sockaddr *) &sa, len, hbuf, sizeof(hbuf),
        NULL, 0, NI_NAMEREQD)) {
        printf("could not resolve hostname\n");
    }
    else {
        printf("host=%s\n", hbuf);
    }

    printf("-- Test 2 --\n");

    he = gethostbyname("www.google.com");
    if (he == NULL) { // do some error checking
        herror("gethostbyname"); // herror(), NOT perror()
        return 1;
    }

    printf("Official name is: %s\n", he->h_name);
    printf("IP address: %s\n", inet_ntoa(*(struct in_addr*)he->h_addr));
    printf("All addresses: ");
    addr_list = (struct in_addr **)he->h_addr_list;
    for(i = 0; addr_list[i] != NULL; i++) {
        printf("%s ", inet_ntoa(*addr_list[i]));
    }
    printf("\n");

    printf("-- Test 3 --\n");
    inet_aton("8.8.8.8", &addr);
    he = gethostbyaddr(&addr, sizeof(addr), AF_INET);
    printf("Host name: %s\n", he->h_name);

    return 0;
}

```

**Comment by rmustacc** [comment]
@damon-atkins What you describe is actually what does happen in python. @sjorge had brought up some of this in illumos. What we found was that the Python implementation in C, always called getaddrinfo() on the name if it was a hostname. So it looked up the hostname to get an address, and then just chooses the first address and passed that back to the underlying system getaddrbyname() function.

**Comment by damon-atkins** [comment]
So why is this not working on Solaris (maybe other platforms we have not tested)?  Seems we have a work around, but not the root cause?

**Comment by rmustacc** [comment]
I can't speak for Solaris, only illumos. In this case it turns up that getaddrinfo() returns an IPv6 address first. The re-entrant versions of gethostbyaddr() never supported IPv6. The non-re-entrant ones will call into getipnodebyaddr(), but really, these are the wrong interfaces to be using in general. They've been deprecated on Linux and the getaddrinfo() and getnameinfo() family was designed to deal with both v4, v6, and handling re-entrancy. For what it's worth, these were never documented to work with IPv6.

I'm not too familiar with the salt level issues, I just was asked to look at the python behavior from an illumos perspective. Reasons this may or may not be an issue on other platforms are a different order of returned addresses or that their re-entrant versions of gethostbyaddr() (which are different on many platforms) have been updated to deal with this despite being deprecated.

**Comment by damon-atkins** [comment]
So this is a bug waiting to happen, because its dependant on the order of  IPv6, IPv4 returned.
For example some platforms take the IPv4 and turn it into automatic IPv6 address as well.  I would say a lot of people most likely do not add the IPv6 address into DNS or host file.

getaddrinfo() and getnameinfo()  Solaris 11 has these functions, and they look the same as Linux.

**Both Linux and Solaris say they order the IP's as per  RFC 3484, Default Address Selection  for  Internet  Protocol version  6 (IPv6).**    So they should be both giving the same results.

Looking at the minion configs, it appears salt does not support IPv4 and IPv6 at the same time.

Seems to me the code needs to lookup the short hostname, which returns a set of IPv4/IPv6 addresses, then it needs to lookup each IP address return and check if it has a result for the IPv4/IPv6 and then check if it starts with the `^short hostname\.`   If it has both a IPv4 and IPv6 hostname it selects the IP protocol Salt is using, if Salt  does not support both at the same time. If none of the FQDN return start with the short hostname ....?

@rmustacc  the idea is a reliable way to get FQDN for the server.  But I am think wider than just Solaris i.e. all platforms/OS.

**Comment by isbm** [comment]
@sjorge Hmm. Tried on Linux: SLE, Leap and Ubuntu, everywhere this works fine:
```
socket.gethostbyaddr(socket.gethostbyname(socket.gethostname()))[0]
```

**Comment by sjorge** [comment]
@isbm Yeah it seems to give the expected results on most platforms I tried it on. (OpenBSD, SmartOS, Solaris 10, FreeBSD and NetBSD)

I just tried it on the later 2 as I just setup a vm with those for status.uptime

**Comment by isbm** [comment]
@sjorge ...but does not work on OSX Sierra and OSX El Capitan. Just use `getfqdn` instead. But lately OSX without Steve Jobs is like a space rocket without a payload. :-(

**Comment by sjorge** [comment]
Oh that's a shame it fails on OS X :/ I guess we learned again that even though the API name is the same... the implementations are not always exact matches 👎

**Comment by isbm** [comment]
@sjorge On Sierra it is even strange. If you call `socket.getfqdn()` then it has a pause of 2-3 seconds...

**Comment by sjorge** [comment]
@isbm do you have AF_INET6 addresses?

I have a dual_stack setup and I noticed that Sierra seems to prefer AF_INET6 over AF_INET more so than previous versions. I also noticed if I am on a network without AF_INET6 addresses that work,... there is general delay in the resolver for a few minutes because it keeps trying AF_INET6 sockets first.

**Comment by isbm** [comment]
@sjorge Well, preferring IPv6 doesn't mean they should suck that much. :laughing:

**Comment by damon-atkins** [comment]
What if someone had in  their DNS like
aname  A 1.1.1.1
anama  A 1.1.1.2
anama  A 1.1.1.3
1 PTR aname-mgt.domain
2 PTR aname-dmz.domain
3 PTR aname.domain

Keeping in mine DNS rotates the results, so the first one will change?
Which one would you get as the FQDN?

See my above comment for a fix.

**Comment by sjorge** [comment]
@damon-atkins yep, that gets messy... as there are also resolver caches to worry about :S

**Comment by sjorge** [comment]
@thatch45 any more progress on this? Or should I create a PR updating the docs marking this as broken on Solaris-like platforms?

**Comment by isbm** [comment]
@sjorge I afraid, go for it. I have to admit, I also did not found enough time to look into that, but we could join The Force. :-)

**Comment by sjorge** [comment]
@isbm I opened PR #38615

Going over the issue again it does seem like: ```socket.gethostbyaddr(socket.gethostbyname(socket.gethostname()))[0]``` would be a valid solution to always get a FQDN on both Solaris (SmartOS/OmniOS retested), FreeBSD and Linux.

**Comment by damon-atkins** [comment]
See https://github.com/saltstack/salt/issues/37027#issuecomment-257074887

**Comment by sjorge** [comment]
This is still a small problem, but both upstream (Illumos) and python do not agree with each other. Python says the API used to query should support both IPv6 and IPv4 like Linux does. Illumos does not agree. 

This still results in incorrect grain data after all the code trying to do the right thing (which was confusing) got taken out. As a workaround I still manually override those grains. Far from perfect though. 

But let’s close this as salt is not really the place and both upstreams are pointing the figure at each other.
